src/CoreWCF.Http/src/CoreWCF/Channels/HttpChannelHelpers.cs

```diff
@@ -121,74 +123,81 @@ private void ThrowMaxReceivedMessageSizeExceeded()
             }
         }
 
-        private async Task<Message> DecodeBufferedMessageAsync(ArraySegment<byte> buffer, Stream inputStream)
+        private async ValueTask<Message> DecodeBufferedMessageAsync(ReadOnlySequence<byte> buffer)
         {
             try
             {
-                // if we're chunked, make sure we've consumed the whole body
-                if (ContentLength == -1 && buffer.Count == _settings.MaxReceivedMessageSize)
+                // if we're chunked, ensure message size is lower than or equal to MaxReceivedMessageSize
+                if (ContentLength == -1 && buffer.Length > _settings.MaxReceivedMessageSize)
```

Comment:
This check needs to move to `ReadBufferedMessageAsync(Stream inputStream)` and check that buffer.Length isn't more the MaxReceivedMessageSize after each call to ReadAsync. The purpose of this is if you only expect small messages to be received, eg less than 1MB, and a malicious client sends you a 1GB request, you want to stop allocating memory reading the request after 1MB. This current implementation would read the entire 1GB and then decide it's over 1MB.

Reply:
this shoudl be ok too now. I restored the ReadBufferedMessageAsync name